examples/virtual_methods/base_button/mod.rs

```diff
@@ -65,24 +65,16 @@ impl<T: BaseButtonImpl> BaseButtonImplExt for T {
     fn parent_sync_method(&self, obj: &BaseButton, extra_text: Option<String>) {
         unsafe {
             let data = Self::type_data();
-            let parent_class = data.as_ref().parent_class() as *mut imp::BaseButtonClass;
-            if let Some(ref f) = (*parent_class).sync_method {
-                f(obj, extra_text)
-            } else {
-                unimplemented!()
-            }
+            let parent_class = &*(data.as_ref().parent_class() as *mut imp::BaseButtonClass);
```

Comment:
I think by making the `TypeData` generic over `T: ObjectSubclass`, we can have `parent_class()` return a properly typed reference and get rid of the `unsafe` code around here.

@vhdirk @Cogitri Does one of you want to give this a try?

Reply:
You mean something like:

```rust
pub struct TypeData<T> where T: ObjectSubclass {
    #[doc(hidden)]
    pub type_: Type,
    #[doc(hidden)]
    pub parent_class: *mut T::Class,
    #[doc(hidden)]
    pub parent_ifaces: Option<HashMap<Type, ffi::gpointer>>,
    #[doc(hidden)]
    pub class_data: Option<HashMap<Type, Box<dyn Any + Send + Sync>>>,
    #[doc(hidden)]
    pub private_offset: isize,
    #[doc(hidden)]
    pub private_imp_offset: isize,
}

```

Reply:
Yes, and `parent_class() -> &T::Class` with a `NULL` check first

Reply:
As said in chat: these generic will propagate upwards to `ObjectSubclass`, where you'd need make the `ObjectSubclassType` constraint generic over an associated type of `ObjectSubclass` itself. I don't think that's possible.